// @polkadot/wasm-crypto/bundle.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hmacSha512 = exports.hmacSha256 = exports.ed25519Verify = exports.ed25519Sign = exports.ed25519KeypairFromSeed = exports.blake2b = exports.bip39Validate = exports.bip39ToSeed = exports.bip39ToMiniSecret = exports.bip39ToEntropy = exports.bip39Generate = void 0;
exports.isReady = isReady;
exports.keccak512 = exports.keccak256 = void 0;
Object.defineProperty(exports, "packageInfo", {
  enumerable: true,
  get: function () {
    return _packageInfo.packageInfo;
  }
});
exports.vrfVerify = exports.vrfSign = exports.twox = exports.sr25519Verify = exports.sr25519Sign = exports.sr25519KeypairFromSeed = exports.sr25519DerivePublicSoft = exports.sr25519DeriveKeypairSoft = exports.sr25519DeriveKeypairHard = exports.sr25519Agree = exports.sha512 = exports.sha256 = exports.scrypt = exports.pbkdf2 = void 0;
exports.waitReady = waitReady;

var _wasmCryptoAsmjs = require("@polkadot/wasm-crypto-asmjs");

var _wasmCryptoWasm = require("@polkadot/wasm-crypto-wasm");

var _bridge = require("./bridge.cjs");

var imports = _interopRequireWildcard(require("./imports.cjs"));

var _packageInfo = require("./packageInfo.cjs");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// Copyright 2019-2021 @polkadot/wasm-crypto authors & contributors
// SPDX-License-Identifier: Apache-2.0
const PTR_VEC = 8;
const wasmPromise = (0, _bridge.initWasm)(_wasmCryptoWasm.wasmBytes, _wasmCryptoAsmjs.asmJsInit, imports).catch(() => null);
const bip39Generate = (0, _bridge.withWasm)(wasm => words => {
  wasm.ext_bip39_generate(PTR_VEC, words);
  return (0, _bridge.resultString)();
});
exports.bip39Generate = bip39Generate;
const bip39ToEntropy = (0, _bridge.withWasm)(wasm => phrase => {
  const [ptr0, len0] = (0, _bridge.allocString)(phrase);
  wasm.ext_bip39_to_entropy(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.bip39ToEntropy = bip39ToEntropy;
const bip39ToMiniSecret = (0, _bridge.withWasm)(wasm => (phrase, password) => {
  const [ptr0, len0] = (0, _bridge.allocString)(phrase);
  const [ptr1, len1] = (0, _bridge.allocString)(password);
  wasm.ext_bip39_to_mini_secret(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.bip39ToMiniSecret = bip39ToMiniSecret;
const bip39ToSeed = (0, _bridge.withWasm)(wasm => (phrase, password) => {
  const [ptr0, len0] = (0, _bridge.allocString)(phrase);
  const [ptr1, len1] = (0, _bridge.allocString)(password);
  wasm.ext_bip39_to_seed(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.bip39ToSeed = bip39ToSeed;
const bip39Validate = (0, _bridge.withWasm)(wasm => phrase => {
  const [ptr0, len0] = (0, _bridge.allocString)(phrase);
  const ret = wasm.ext_bip39_validate(ptr0, len0);
  return ret !== 0;
});
exports.bip39Validate = bip39Validate;
const ed25519KeypairFromSeed = (0, _bridge.withWasm)(wasm => seed => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(seed);
  wasm.ext_ed_from_seed(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.ed25519KeypairFromSeed = ed25519KeypairFromSeed;
const ed25519Sign = (0, _bridge.withWasm)(wasm => (pubkey, seckey, message) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pubkey);
  const [ptr1, len1] = (0, _bridge.allocU8a)(seckey);
  const [ptr2, len2] = (0, _bridge.allocU8a)(message);
  wasm.ext_ed_sign(PTR_VEC, ptr0, len0, ptr1, len1, ptr2, len2);
  return (0, _bridge.resultU8a)();
});
exports.ed25519Sign = ed25519Sign;
const ed25519Verify = (0, _bridge.withWasm)(wasm => (signature, message, pubkey) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(signature);
  const [ptr1, len1] = (0, _bridge.allocU8a)(message);
  const [ptr2, len2] = (0, _bridge.allocU8a)(pubkey);
  const ret = wasm.ext_ed_verify(ptr0, len0, ptr1, len1, ptr2, len2);
  return ret !== 0;
});
exports.ed25519Verify = ed25519Verify;
const sr25519DeriveKeypairHard = (0, _bridge.withWasm)(wasm => (pair, cc) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pair);
  const [ptr1, len1] = (0, _bridge.allocU8a)(cc);
  wasm.ext_sr_derive_keypair_hard(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.sr25519DeriveKeypairHard = sr25519DeriveKeypairHard;
const sr25519DeriveKeypairSoft = (0, _bridge.withWasm)(wasm => (pair, cc) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pair);
  const [ptr1, len1] = (0, _bridge.allocU8a)(cc);
  wasm.ext_sr_derive_keypair_soft(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.sr25519DeriveKeypairSoft = sr25519DeriveKeypairSoft;
const sr25519DerivePublicSoft = (0, _bridge.withWasm)(wasm => (pubkey, cc) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pubkey);
  const [ptr1, len1] = (0, _bridge.allocU8a)(cc);
  wasm.ext_sr_derive_public_soft(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.sr25519DerivePublicSoft = sr25519DerivePublicSoft;
const sr25519KeypairFromSeed = (0, _bridge.withWasm)(wasm => seed => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(seed);
  wasm.ext_sr_from_seed(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.sr25519KeypairFromSeed = sr25519KeypairFromSeed;
const sr25519Sign = (0, _bridge.withWasm)(wasm => (pubkey, secret, message) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pubkey);
  const [ptr1, len1] = (0, _bridge.allocU8a)(secret);
  const [ptr2, len2] = (0, _bridge.allocU8a)(message);
  wasm.ext_sr_sign(PTR_VEC, ptr0, len0, ptr1, len1, ptr2, len2);
  return (0, _bridge.resultU8a)();
});
exports.sr25519Sign = sr25519Sign;
const sr25519Verify = (0, _bridge.withWasm)(wasm => (signature, message, pubkey) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(signature);
  const [ptr1, len1] = (0, _bridge.allocU8a)(message);
  const [ptr2, len2] = (0, _bridge.allocU8a)(pubkey);
  const ret = wasm.ext_sr_verify(ptr0, len0, ptr1, len1, ptr2, len2);
  return ret !== 0;
});
exports.sr25519Verify = sr25519Verify;
const sr25519Agree = (0, _bridge.withWasm)(wasm => (pubkey, secret) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pubkey);
  const [ptr1, len1] = (0, _bridge.allocU8a)(secret);
  wasm.ext_sr_agree(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.sr25519Agree = sr25519Agree;
const vrfSign = (0, _bridge.withWasm)(wasm => (secret, context, message, extra) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(secret);
  const [ptr1, len1] = (0, _bridge.allocU8a)(context);
  const [ptr2, len2] = (0, _bridge.allocU8a)(message);
  const [ptr3, len3] = (0, _bridge.allocU8a)(extra);
  wasm.ext_vrf_sign(PTR_VEC, ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3);
  return (0, _bridge.resultU8a)();
});
exports.vrfSign = vrfSign;
const vrfVerify = (0, _bridge.withWasm)(wasm => (pubkey, context, message, extra, outAndProof) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(pubkey);
  const [ptr1, len1] = (0, _bridge.allocU8a)(context);
  const [ptr2, len2] = (0, _bridge.allocU8a)(message);
  const [ptr3, len3] = (0, _bridge.allocU8a)(extra);
  const [ptr4, len4] = (0, _bridge.allocU8a)(outAndProof);
  const ret = wasm.ext_vrf_verify(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, ptr4, len4);
  return ret !== 0;
});
exports.vrfVerify = vrfVerify;
const blake2b = (0, _bridge.withWasm)(wasm => (data, key, size) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  const [ptr1, len1] = (0, _bridge.allocU8a)(key);
  wasm.ext_blake2b(PTR_VEC, ptr0, len0, ptr1, len1, size);
  return (0, _bridge.resultU8a)();
});
exports.blake2b = blake2b;
const hmacSha256 = (0, _bridge.withWasm)(wasm => (key, data) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(key);
  const [ptr1, len1] = (0, _bridge.allocU8a)(data);
  wasm.ext_hmac_sha256(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.hmacSha256 = hmacSha256;
const hmacSha512 = (0, _bridge.withWasm)(wasm => (key, data) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(key);
  const [ptr1, len1] = (0, _bridge.allocU8a)(data);
  wasm.ext_hmac_sha512(PTR_VEC, ptr0, len0, ptr1, len1);
  return (0, _bridge.resultU8a)();
});
exports.hmacSha512 = hmacSha512;
const keccak256 = (0, _bridge.withWasm)(wasm => data => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  wasm.ext_keccak256(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.keccak256 = keccak256;
const keccak512 = (0, _bridge.withWasm)(wasm => data => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  wasm.ext_keccak512(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.keccak512 = keccak512;
const pbkdf2 = (0, _bridge.withWasm)(wasm => (data, salt, rounds) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  const [ptr1, len1] = (0, _bridge.allocU8a)(salt);
  wasm.ext_pbkdf2(PTR_VEC, ptr0, len0, ptr1, len1, rounds);
  return (0, _bridge.resultU8a)();
});
exports.pbkdf2 = pbkdf2;
const scrypt = (0, _bridge.withWasm)(wasm => (password, salt, log2n, r, p) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(password);
  const [ptr1, len1] = (0, _bridge.allocU8a)(salt);
  wasm.ext_scrypt(PTR_VEC, ptr0, len0, ptr1, len1, log2n, r, p);
  return (0, _bridge.resultU8a)();
});
exports.scrypt = scrypt;
const sha256 = (0, _bridge.withWasm)(wasm => data => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  wasm.ext_sha256(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.sha256 = sha256;
const sha512 = (0, _bridge.withWasm)(wasm => data => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  wasm.ext_sha512(PTR_VEC, ptr0, len0);
  return (0, _bridge.resultU8a)();
});
exports.sha512 = sha512;
const twox = (0, _bridge.withWasm)(wasm => (data, rounds) => {
  const [ptr0, len0] = (0, _bridge.allocU8a)(data);
  wasm.ext_twox(PTR_VEC, ptr0, len0, rounds);
  return (0, _bridge.resultU8a)();
});
exports.twox = twox;

function isReady() {
  return !!(0, _bridge.getWasm)();
}

function waitReady() {
  return wasmPromise.then(() => isReady());
}
